Ian management team comprehensive requirement

ianhomew opened the pull request:
Fixed problems in 管理組.xlsx

nick66551 commented on app/Http/Controllers/MaintenanceController.php:
```diff
@@ -314,4 +302,18 @@ private function createLandlordPayment($maintenance): void
+    private function updateRoomStatusIfNeeded($room)
+    {
+        $id = \request()->input('tenant_contract_id');
+        $type = \request()->input('incident_type');
```
If parameters of this function been got by request(), it may be hard to do unit tests in the future?

ianhomew replied:
changed to called by params.

nick66551 commented on resources/views/maintenances/index.blade.php:
```diff
@@ -20,161 +20,149 @@
+                                                    <td>
+                                                        <a class="btn btn-success btn-xs" href="{{ route( 'maintenances.show', $maintenance['id']) }}?with=tenant;room;documents">查看</a>
+                                                        <a class="btn btn-primary btn-xs" href="{{ route( 'maintenances.edit', $maintenance['id']) }}">編輯</a>
+                                                        <a class="btn btn-danger btn-xs jquery-postback" data-method="delete" href="{{ route('maintenances.show', $maintenance['id']) }}">刪除</a>
```
'刪除' button should be deleted?

ianhomew replied:
deleted

nick66551 commented on app/Services/ScheduleService.php:
```diff
@@ -140,18 +140,24 @@ public function notifyTenantContractDueInTwoMonths()
-            ->where('updated_at', '<=', $limitDatetime)
+        $maintenances = Maintenance::where('status', '!=', '案件完成')
+            ->orWhere(function ($query) {
+                $query->where('status', '!=', '已取消');
```
would `whereNotIn('status', ['案件完成', '已取消'])` be better and meet the logic  ?

ianhomew replied:
changed

nick66551 commented on app/Services/ScheduleService.php:
```diff
@@ -140,18 +140,24 @@ public function notifyTenantContractDueInTwoMonths()
             $commissioner->notify(
                 new TextNotify(
-                    "維修清潔單號：{$maintenance->id} 狀態超過 {$notifyRequiredDays} 未更新，煩請抽空查看。"
+                    "維修清潔單號：{$maintenance->id} 狀態超過 {$notifyRequiredDays} 天未更新，煩請抽空查看。 $url"
```
The $url would display like `<a href="/maintenances/1">cc-house.cc/maintenances/1</a>` and can be clicked directly?

ianhomew replied:
added the <a> tag for clicking

nick66551 commented on app/Services/ScheduleService.php:
```diff
@@ -1,6 +1,7 @@
 <?php
```
You mention `modified maintenance notification` on commit message, but doesn't see the change of this commit?

ianhomew replied:
search 'notifyMaintenanceStatus' and u'll see // 超過『預計處理日期』 時，且『狀態』不是『案件完成』和 『已取消』才跳通知

nick66551 commented on app/Services/RoomService.php:
```diff
@@ -61,9 +71,39 @@ public static function update($room, $data, $appliances)
+    private static function notifyMaintenanceBuilder(Room $room, array $data)
+    {
+        if ($data['room_status'] !== $room->room_status) {
+            $tenantContractIds = collect($room->tenantContracts)->map->id->toArray();
```
tried `$room->tenantContracts->pluck('id')->toArray()` ?

ianhomew replied:
nice

nick66551 commented:
Have you finished function below? :
1. `清潔維修管理` - `查看` - 照片會顯示在下方
2. `鑰匙管理` - `鑰匙紀錄資料新增` 
-  增加備註欄位(非必填)
- 『狀態』，"使用中"改"持有中"、"已完成"改"已歸還"
3. `鑰匙管理` -  `鑰匙紀錄資料刪除` - 取消刪除功能

ianhomew commented on app/Http/Controllers/TenantController.php:
```diff
@@ -231,31 +231,22 @@ private function updateRelatedPeople(Tenant $tenant, array $relatedPeople)
+            ->contactInfos()
+            ->where('contactable_type', Tenant::class)
+            ->delete();
 
```
I revert the code to last version and changed the contactable_type

ianhomew commented on app/Http/Controllers/TenantContractController.php:
```diff
@@ -281,7 +287,8 @@ public function update(Request $request, TenantContract $tenantContract)
             ],
             'invoice_collection_number' => 'required|max:255',
-            'commissioner_id' => 'exists:users,id'
+            'commissioner_id' => 'exists:users,id',
```
all 220_* in store and update function will changed from required to nullable

ianhomew commented:
> Have you finished function below? :
> 
> 1. `清潔維修管理` - `查看` - 照片會顯示在下方
> 2. `鑰匙管理` - `鑰匙紀錄資料新增`
> 
> * 增加備註欄位(非必填)
> * 『狀態』，"使用中"改"持有中"、"已完成"改"已歸還"
> 
> 1. `鑰匙管理` -  `鑰匙紀錄資料刪除` - 取消刪除功能


1. 
![image](https://user-images.githubusercontent.com/16000743/66095833-9c9a7200-e5cb-11e9-9159-bea133121f5f.png)

2. added

3. removed delete button on key request page.
